File: utk_curio/backend/app/packages/node_appearance.py
"""THE shared node-appearance utility (memo dev/89 §3).

One validation/normalization/derivation truth for the per-node
``appearance.backgroundColor`` used by: build-request parsing
(``build_models``), proposal minting and Apply (``agents.services``), and —
mirrored constant-for-constant — the frontend behavior's color logic. No
caller re-implements color rules; forking this logic is the bug the memo
forbids.

Contract (dev/89 §3 "Node Researcher DOD profile"):

* Named palette — ``yellow`` (default), ``pink``, ``blue``, ``green``,
  ``orange``, ``lavender`` — mapped centrally to design-token hex values.
* Custom colors: a normalized six-digit ``#RRGGBB`` only. Shorthand hex,
  alpha channels, ``rgb()``/gradients/``url()``/CSS expressions, and
  whitespace tricks are rejected loudly.
* Accessibility is part of validity: a background no foreground can reach
  WCAG AA (4.5:1) against is REJECTED at normalization time, not rendered
  unreadable (dev/89 §6).
* Legacy tolerance lives in ONE place: :func:`resolve_background` maps a
  missing/invalid stored value to the default yellow for rendering — new
  writes always go through :func:`normalize_appearance`, which raises.

Derivations (:func:`derived_colors`) hand the behavior its foreground,
muted-foreground, border, link, and focus colors — the model and the stored
spec never carry raw text/border CSS (dev/89 §3).
"""

from __future__ import annotations

import re
from typing import Any, Mapping

#: Central palette — design tokens, mirrored by the frontend utility.
NAMED_COLORS: dict[str, str] = {
    "yellow": "#fef3c0",
    "pink": "#fbd3e0",
    "blue": "#cfe8f7",
    "green": "#d5f0d1",
    "orange": "#ffddc0",
    "lavender": "#e4dcf7",
}

DEFAULT_COLOR_NAME = "yellow"
DEFAULT_BACKGROUND = NAMED_COLORS[DEFAULT_COLOR_NAME]

# WCAG AA for normal text.
MIN_CONTRAST = 4.5

# Foreground candidates: the ink pair the behavior renders with.
_DARK_FOREGROUND = "#1f2430"
_LIGHT_FOREGROUND = "#ffffff"
# Link candidates (light-bg blue / dark-bg blue), falling back to the
# foreground when neither clears AA against the background.
_LINK_ON_LIGHT = "#1d4ed8"
_LINK_ON_DARK = "#93c5fd"

_HEX_RE = re.compile(r"^#[0-9a-fA-F]{6}$")


class AppearanceError(ValueError):
    """Raised when a requested appearance value is invalid or inaccessible."""


def _rgb(hex_color: str) -> tuple[int, int, int]:
    return (int(hex_color[1:3], 16), int(hex_color[3:5], 16), int(hex_color[5:7], 16))


def _hex(rgb: tuple[int, int, int]) -> str:
    return "#" + "".join(f"{max(0, min(255, round(c))):02x}" for c in rgb)


def _luminance(hex_color: str) -> float:
    def _lin(channel: int) -> float:
        s = channel / 255.0
        return s / 12.92 if s <= 0.04045 else ((s + 0.055) / 1.055) ** 2.4

    r, g, b = _rgb(hex_color)
    return 0.2126 * _lin(r) + 0.7152 * _lin(g) + 0.0722 * _lin(b)


def contrast_ratio(hex_a: str, hex_b: str) -> float:
    """WCAG contrast ratio between two ``#RRGGBB`` colors."""
    la, lb = _luminance(hex_a), _luminance(hex_b)
    lighter, darker = max(la, lb), min(la, lb)
    return (lighter + 0.05) / (darker + 0.05)


def _best_foreground(background: str) -> tuple[str, float]:
    dark = contrast_ratio(background, _DARK_FOREGROUND)
    light = contrast_ratio(background, _LIGHT_FOREGROUND)
    return (_DARK_FOREGROUND, dark) if dark >= light else (_LIGHT_FOREGROUND, light)


def normalize_background(value: Any) -> str:
    """Normalize one requested background to a lowercase ``#RRGGBB``.

    Accepts a palette name (case-insensitive) or a six-digit hex. Everything
    else — shorthand hex, alpha, CSS functions/keywords, embedded whitespace
    — raises :class:`AppearanceError`, as does a hex no foreground can read
    at WCAG AA.
    """
    if not isinstance(value, str) or not value:
        raise AppearanceError(
            "backgroundColor must be a palette name "
            f"({', '.join(sorted(NAMED_COLORS))}) or a six-digit #RRGGBB hex"
        )
    candidate = value.strip()
    if candidate != value or any(c.isspace() for c in candidate):
        raise AppearanceError(f"backgroundColor {value!r} contains whitespace — refused")
    named = NAMED_COLORS.get(candidate.lower())
    if named is not None:
        return named
    if not _HEX_RE.match(candidate):
        raise AppearanceError(
            f"backgroundColor {value!r} is not a palette name or six-digit "
            "#RRGGBB hex (shorthand hex, alpha channels, and CSS expressions "
            "are refused)"
        )
    normalized = candidate.lower()
    _, best = _best_foreground(normalized)
    if best < MIN_CONTRAST:
        raise AppearanceError(
            f"backgroundColor {value!r} cannot produce WCAG AA text contrast "
            f"(best foreground reaches {best:.2f}:1, needs {MIN_CONTRAST}:1) — "
            "choose a lighter or darker color"
        )
    return normalized


def normalize_appearance(raw: Any) -> dict[str, str] | None:
    """Validate one appearance object: ``None`` stays ``None``; otherwise
    exactly ``{"backgroundColor": <name-or-hex>}`` normalized to hex."""
    if raw is None:
        return None
    if not isinstance(raw, Mapping):
        raise AppearanceError("appearance must be an object")
    unknown = set(raw) - {"backgroundColor"}
    if unknown:
        raise AppearanceError(
            f"appearance has unknown keys {sorted(unknown)}; only "
            "backgroundColor is supported"
        )
    if "backgroundColor" not in raw:
        raise AppearanceError("appearance requires backgroundColor")
    return {"backgroundColor": normalize_background(raw["backgroundColor"])}


def resolve_background(stored: Any) -> str:
    """Render-path tolerance: a stored value that is missing or no longer
    valid falls back to the default yellow — QUIETLY, because legacy specs
    are data to render, not requests to refuse (dev/89 §6)."""
    try:
        return normalize_background(stored)
    except AppearanceError:
        return DEFAULT_BACKGROUND


def _shade(hex_color: str, factor: float) -> str:
    """factor < 1 darkens toward black; factor > 1 lightens toward white."""
    r, g, b = _rgb(hex_color)
    if factor <= 1:
        return _hex((r * factor, g * factor, b * factor))
    t = min(1.0, factor - 1.0)
    return _hex((r + (255 - r) * t, g + (255 - g) * t, b + (255 - b) * t))


def derived_colors(background: Any) -> dict[str, str]:
    """Foreground, muted text, border, link, and focus colors for one
    background — every pair AA-safe by construction. Computed at render
    time, never persisted (dev/89 §4)."""
    bg = resolve_background(background)
    foreground, _ = _best_foreground(bg)
    dark_ink = foreground == _DARK_FOREGROUND
    link = _LINK_ON_LIGHT if dark_ink else _LINK_ON_DARK
    if contrast_ratio(bg, link) < MIN_CONTRAST:
        link = foreground
    # Muted text: pull the ink 25% toward the background; step back toward
    # full ink until it clears AA (always terminates at the foreground).
    fr, fg_, fb = _rgb(foreground)
    br, bgc, bb = _rgb(bg)
    for mix in (0.25, 0.15, 0.0):
        muted = _hex((fr + (br - fr) * mix, fg_ + (bgc - fg_) * mix, fb + (bb - fb) * mix))
        if contrast_ratio(bg, muted) >= MIN_CONTRAST:
            break
    border = _shade(bg, 0.82 if dark_ink else 1.35)
    return {
        "background": bg,
        "foreground": foreground,
        "mutedForeground": muted,
        "border": border,
        "link": link,
        "focus": foreground,
    }

File: utk_curio/frontend/urban-workflows/src/utils/nodeAppearance.ts
/**
 * THE frontend node-appearance utility (memo dev/89 §3) — the
 * constant-for-constant mirror of the backend's
 * `utk_curio/backend/app/packages/node_appearance.py`. One validation /
 * normalization / derivation truth per layer; behaviors and controls never
 * re-implement color rules.
 *
 * Contract: a named palette (yellow default, pink, blue, green, orange,
 * lavender) mapped centrally to design-token hexes, plus six-digit `#RRGGBB`
 * custom colors only. Shorthand hex, alpha, CSS expressions, and whitespace
 * are refused; a background no foreground can read at WCAG AA (4.5:1) is
 * invalid. Legacy tolerance lives ONLY in {@link resolveBackground}: stored
 * junk renders as the default yellow, never as unreadable ink.
 */

export const NAMED_COLORS: Record<string, string> = {
  yellow: "#fef3c0",
  pink: "#fbd3e0",
  blue: "#cfe8f7",
  green: "#d5f0d1",
  orange: "#ffddc0",
  lavender: "#e4dcf7",
};

export const DEFAULT_COLOR_NAME = "yellow";
export const DEFAULT_BACKGROUND = NAMED_COLORS[DEFAULT_COLOR_NAME];

/** WCAG AA for normal text. */
export const MIN_CONTRAST = 4.5;

const DARK_FOREGROUND = "#1f2430";
const LIGHT_FOREGROUND = "#ffffff";
const LINK_ON_LIGHT = "#1d4ed8";
const LINK_ON_DARK = "#93c5fd";

const HEX_RE = /^#[0-9a-fA-F]{6}$/;

export type NodeAppearance = { backgroundColor: string };

export type DerivedColors = {
  background: string;
  foreground: string;
  mutedForeground: string;
  border: string;
  link: string;
  focus: string;
};

function rgb(hex: string): [number, number, number] {
  return [
    parseInt(hex.slice(1, 3), 16),
    parseInt(hex.slice(3, 5), 16),
    parseInt(hex.slice(5, 7), 16),
  ];
}

function toHex(channels: [number, number, number]): string {
  return (
    "#" +
    channels
      .map((c) => Math.max(0, Math.min(255, Math.round(c))).toString(16).padStart(2, "0"))
      .join("")
  );
}

function luminance(hex: string): number {
  const lin = (channel: number) => {
    const s = channel / 255;
    return s <= 0.04045 ? s / 12.92 : Math.pow((s + 0.055) / 1.055, 2.4);
  };
  const [r, g, b] = rgb(hex);
  return 0.2126 * lin(r) + 0.7152 * lin(g) + 0.0722 * lin(b);
}

/** WCAG contrast ratio between two `#RRGGBB` colors. */
export function contrastRatio(hexA: string, hexB: string): number {
  const la = luminance(hexA);
  const lb = luminance(hexB);
  const lighter = Math.max(la, lb);
  const darker = Math.min(la, lb);
  return (lighter + 0.05) / (darker + 0.05);
}

function bestForeground(background: string): { foreground: string; ratio: number } {
  const dark = contrastRatio(background, DARK_FOREGROUND);
  const light = contrastRatio(background, LIGHT_FOREGROUND);
  return dark >= light
    ? { foreground: DARK_FOREGROUND, ratio: dark }
    : { foreground: LIGHT_FOREGROUND, ratio: light };
}

/**
 * Normalize one requested background to a lowercase `#RRGGBB`, or return the
 * refusal reason. Mirrors the backend: palette names (case-insensitive) or
 * six-digit hex only; whitespace, shorthand, alpha, CSS expressions, and
 * AA-unreachable colors are refused.
 */
export function normalizeBackground(
  value: unknown,
): { ok: true; value: string } | { ok: false; reason: string } {
  if (typeof value !== "string" || !value) {
    return {
      ok: false,
      reason: `Pick a palette color (${Object.keys(NAMED_COLORS).sort().join(", ")}) or a six-digit #RRGGBB hex.`,
    };
  }
  if (/\s/.test(value)) {
    return { ok: false, reason: "The color must not contain whitespace." };
  }
  const named = NAMED_COLORS[value.toLowerCase()];
  if (named !== undefined) return { ok: true, value: named };
  if (!HEX_RE.test(value)) {
    return {
      ok: false,
      reason:
        "Not a palette name or six-digit #RRGGBB hex (shorthand hex, alpha, and CSS expressions are not supported).",
    };
  }
  const normalized = value.toLowerCase();
  const { ratio } = bestForeground(normalized);
  if (ratio < MIN_CONTRAST) {
    return {
      ok: false,
      reason: `That color can't show readable text (best contrast ${ratio.toFixed(2)}:1, needs ${MIN_CONTRAST}:1) — pick a lighter or darker color.`,
    };
  }
  return { ok: true, value: normalized };
}

/**
 * Render-path tolerance (the ONLY quiet fallback): a stored value that is
 * missing or no longer valid renders as the default yellow.
 */
export function resolveBackground(stored: unknown): string {
  const result = normalizeBackground(stored);
  return result.ok ? result.value : DEFAULT_BACKGROUND;
}

function shade(hex: string, factor: number): string {
  const [r, g, b] = rgb(hex);
  if (factor <= 1) return toHex([r * factor, g * factor, b * factor]);
  const t = Math.min(1, factor - 1);
  return toHex([r + (255 - r) * t, g + (255 - g) * t, b + (255 - b) * t]);
}

/**
 * Foreground, muted text, border, link, and focus colors for one background —
 * every pair AA-safe by construction (mirrors the backend derivation).
 * Computed at render time, never persisted.
 */
export function derivedColors(background: unknown): DerivedColors {
  const bg = resolveBackground(background);
  const { foreground } = bestForeground(bg);
  const darkInk = foreground === DARK_FOREGROUND;
  let link = darkInk ? LINK_ON_LIGHT : LINK_ON_DARK;
  if (contrastRatio(bg, link) < MIN_CONTRAST) link = foreground;
  const [fr, fg, fb] = rgb(foreground);
  const [br, bgc, bb] = rgb(bg);
  let muted = foreground;
  for (const mix of [0.25, 0.15, 0]) {
    const candidate = toHex([
      fr + (br - fr) * mix,
      fg + (bgc - fg) * mix,
      fb + (bb - fb) * mix,
    ]);
    if (contrastRatio(bg, candidate) >= MIN_CONTRAST) {
      muted = candidate;
      break;
    }
  }
  return {
    background: bg,
    foreground,
    mutedForeground: muted,
    border: shade(bg, darkInk ? 0.82 : 1.35),
    link,
    focus: foreground,
  };
}
